[sparsity] Add m-out-of-n support in the WeightNormSparsifier

z-a-f opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #65295
* #65296
* #65294
* #65293
* #65292

The m-out-of-n is implemented as follows:

1. Compute the blocks that need to be sparsified using the weight-norm criterion
2. Within each block below the threshold find the smallest absolute value elements
3. Zero out only the smallest values within each block

m-out-of-n describes sparsification scheme where in a block with "n" elements, only "m" of them would be zeroed-out.
Block sparsity, with the whole block being all zeros, is a special case of m-out-n: If m==n, the whole block is reset.

This echoes the implementation described in the https://github.com/pytorch/pytorch/issues/59835,
as well as meets the support of the nVidia cusparselt requirements.
To support the CUDA sparsity (2/4), one would need to set the sparsity_level to 1.0.
That translates to all blocks of shape 1x4 within a tensor will sprasify with 2-out-4 scheme.

Differential Revision: [D31186828](https://our.internmc.facebook.com/intern/diff/D31186828)

Commit:
[sparsity] Add m-out-of-n support in the WeightNormSparsifier

The m-out-of-n is implemented as follows:

1. Compute the blocks that need to be sparsified using the weight-norm criterion
2. Within each block below the threshold find the smallest absolute value elements
3. Zero out only the smallest values within each block

m-out-of-n describes sparsification scheme where in a block with "n" elements, only "m" of them would be zeroed-out.
Block sparsity, with the whole block being all zeros, is a special case of m-out-n: If m==n, the whole block is reset.

This echoes the implementation described in the https://github.com/pytorch/pytorch/issues/59835,
as well as meets the support of the nVidia cusparselt requirements.
To support the CUDA sparsity (2/4), one would need to set the sparsity_level to 1.0.
That translates to all blocks of shape 1x4 within a tensor will sprasify with 2-out-4 scheme.

[ghstack-poisoned]

Commit:
Update on "[sparsity] Add m-out-of-n support in the WeightNormSparsifier"


The m-out-of-n is implemented as follows:

1. Compute the blocks that need to be sparsified using the weight-norm criterion
2. Within each block below the threshold find the smallest absolute value elements
3. Zero out only the smallest values within each block

m-out-of-n describes sparsification scheme where in a block with "n" elements, only "m" of them would be zeroed-out.
Block sparsity, with the whole block being all zeros, is a special case of m-out-n: If m==n, the whole block is reset.

This echoes the implementation described in the https://github.com/pytorch/pytorch/issues/59835,
as well as meets the support of the nVidia cusparselt requirements.
To support the CUDA sparsity (2/4), one would need to set the sparsity_level to 1.0.
That translates to all blocks of shape 1x4 within a tensor will sprasify with 2-out-4 scheme.

[ghstack-poisoned]

jerryzh168 commented on torch/ao/sparsity/sparsifier/weight_norm_sparsifier.py:
```diff
@@ -54,13 +54,22 @@ def __init__(self,
+        if zeros_per_block > values_per_block:
+            raise ValueError("Number of zeros per block cannot be more than "
+                             "the total number of elements in that block.")
+        elif sparsity_level <= 0.0:
```
nit: probably just `if` since this check is not too relevant to the previous one?
I'm assuming sparsity_level is a percentage of sparsity and when it is <= 0 it means it's not sparsified and when it is >= 1.0 the tensor should be fully sparsified? could you add this to the docstring above?

also does the concept of sparsity_level and values_per_block overlap?

vkuzo replied:
qq, is there a difference between this line and L70?  From reading the code, seems like if sparsity_level <= 0.0 then L63 would return and L70 would not be reached...did I get that right?

z-a-f replied:
> also does the concept of sparsity_level and values_per_block overlap?

Not directly. We define the `sparsity_level` as "ratio of blocks that are sparse". However, `values_per_block` defines the number of elements in a block (basically shape[0] * shape[1] * ...). The latter is then used to compute the sparsity level within each block. So, for example, if `zeros_per_block = 2` and `values_per_block = 4`, that would mean that block is expected to be 50% sparse. I will have a brief about the API early next week

z-a-f replied:
> qq, is there a difference between this line and L70? From reading the code, seems like if sparsity_level <= 0.0 then L63 would return and L70 would not be reached...did I get that right?

Oh yea -- good catch, there is some code duplication going on here. Will recheck it

Commit:
Update on "[sparsity] Add m-out-of-n support in the WeightNormSparsifier"


The m-out-of-n is implemented as follows:

1. Compute the blocks that need to be sparsified using the weight-norm criterion
2. Within each block below the threshold find the smallest absolute value elements
3. Zero out only the smallest values within each block

m-out-of-n describes sparsification scheme where in a block with "n" elements, only "m" of them would be zeroed-out.
Block sparsity, with the whole block being all zeros, is a special case of m-out-n: If m==n, the whole block is reset.

This echoes the implementation described in the https://github.com/pytorch/pytorch/issues/59835,
as well as meets the support of the nVidia cusparselt requirements.
To support the CUDA sparsity (2/4), one would need to set the sparsity_level to 1.0.
That translates to all blocks of shape 1x4 within a tensor will sprasify with 2-out-4 scheme.

[ghstack-poisoned]

z-a-f commented:
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31186828).

Commit:
Update on "[sparsity] Add m-out-of-n support in the WeightNormSparsifier"


The m-out-of-n is implemented as follows:

1. Compute the blocks that need to be sparsified using the weight-norm criterion
2. Within each block below the threshold find the smallest absolute value elements
3. Zero out only the smallest values within each block

m-out-of-n describes sparsification scheme where in a block with "n" elements, only "m" of them would be zeroed-out.
Block sparsity, with the whole block being all zeros, is a special case of m-out-n: If m==n, the whole block is reset.

This echoes the implementation described in the https://github.com/pytorch/pytorch/issues/59835,
as well as meets the support of the nVidia cusparselt requirements.
To support the CUDA sparsity (2/4), one would need to set the sparsity_level to 1.0.
That translates to all blocks of shape 1x4 within a tensor will sprasify with 2-out-4 scheme.

Differential Revision: [D31186828](https://our.internmc.facebook.com/intern/diff/D31186828)

[ghstack-poisoned]

z-a-f commented:
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31186828).

Commit:
Update on "[sparsity] Add m-out-of-n support in the WeightNormSparsifier"


The m-out-of-n is implemented as follows:

1. Compute the blocks that need to be sparsified using the weight-norm criterion
2. Within each block below the threshold find the smallest absolute value elements
3. Zero out only the smallest values within each block

m-out-of-n describes sparsification scheme where in a block with "n" elements, only "m" of them would be zeroed-out.
Block sparsity, with the whole block being all zeros, is a special case of m-out-n: If m==n, the whole block is reset.

This echoes the implementation described in the https://github.com/pytorch/pytorch/issues/59835,
as well as meets the support of the nVidia cusparselt requirements.
To support the CUDA sparsity (2/4), one would need to set the sparsity_level to 1.0.
That translates to all blocks of shape 1x4 within a tensor will sprasify with 2-out-4 scheme.

Differential Revision: [D31186828](https://our.internmc.facebook.com/intern/diff/D31186828)

[ghstack-poisoned]

Commit:
Update on "[sparsity] Add m-out-of-n support in the WeightNormSparsifier"


The m-out-of-n is implemented as follows:

1. Compute the blocks that need to be sparsified using the weight-norm criterion
2. Within each block below the threshold find the smallest absolute value elements
3. Zero out only the smallest values within each block

m-out-of-n describes sparsification scheme where in a block with "n" elements, only "m" of them would be zeroed-out.
Block sparsity, with the whole block being all zeros, is a special case of m-out-n: If m==n, the whole block is reset.

This echoes the implementation described in the https://github.com/pytorch/pytorch/issues/59835,
as well as meets the support of the nVidia cusparselt requirements.
To support the CUDA sparsity (2/4), one would need to set the sparsity_level to 1.0.
That translates to all blocks of shape 1x4 within a tensor will sprasify with 2-out-4 scheme.

Differential Revision: [D31186828](https://our.internmc.facebook.com/intern/diff/D31186828)

[ghstack-poisoned]

z-a-f commented:
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31186828).

z-a-f commented:
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31186828).

Commit:
Update on "[sparsity] Add m-out-of-n support in the WeightNormSparsifier"


The m-out-of-n is implemented as follows:

1. Compute the blocks that need to be sparsified using the weight-norm criterion
2. Within each block below the threshold find the smallest absolute value elements
3. Zero out only the smallest values within each block

m-out-of-n describes sparsification scheme where in a block with "n" elements, only "m" of them would be zeroed-out.
Block sparsity, with the whole block being all zeros, is a special case of m-out-n: If m==n, the whole block is reset.

This echoes the implementation described in the https://github.com/pytorch/pytorch/issues/59835,
as well as meets the support of the nVidia cusparselt requirements.
To support the CUDA sparsity (2/4), one would need to set the sparsity_level to 1.0.
That translates to all blocks of shape 1x4 within a tensor will sprasify with 2-out-4 scheme.

Differential Revision: [D31186828](https://our.internmc.facebook.com/intern/diff/D31186828)

[ghstack-poisoned]

vkuzo commented on test/ao/sparsity/test_sparsifier.py:
```diff
@@ -220,3 +240,57 @@ def test_mask_squash(self):
             assert not is_parametrized(module, 'weight')
             assert not hasattr(module, 'mask')
+
+    def test_sparsity_levels(self):
```
optional: might be good to have at least one test case which tests the behavior directly (hardcode a tensor, apply m-n sparsity, verify individual values).  The current coverage is good but might not catch bugs with sorting, etc.

z-a-f commented:
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31186828).
